### DISCUSSION TITLE
Annotation Plugin

### DISCUSSION THREAD

**Discussion by nadar** [ueberdosis/tiptap#1445] [General Questions]
Hi Everyone

We have made an issue a while ago regarding annotation plugin. We found that in the [experimental section](https://www.tiptap.dev/experiments/collaboration-annotation) and where able to get the code. Since this is experiment, and not released we decided to sponsor the project (`Freelancer/Startup` tier). As we are aware we can not speed up certain things i am looking for help in this thread. If we can fix our issue i would gratefully send a PR for documentations, as it might help others.

If someone else looking for the same, or might help me on this, here you can find the experimental tiptap annotation code: https://github.com/ueberdosis/tiptap/tree/main/docs/src/demos/Experiments/CollaborationAnnotation

We think there is a small part "missing" or not documented to make it available for "real applications":

+ It should be possible to load annotations from outside into the given positions (assuming you like to store the annotations in a database, which seems to be a very common use case :smiley:)
+ This means we would need to access all that information for either ydoc and/or the annotation plugin itself.

Lets assume the following code, which utilizes the annotation plugin. I removed non essential parts to make it more easy to read:

```js
export default {
  data:() => ({
    editor: null,
    ydoc: null,
    map: null,
    comments: [], // holds only the commentr for the given words selected
  }),
  mounted() {
    this.ydoc = new Y.Doc()
    this.map = this.ydoc.getMap('myydocmap')
    this.map.observe(() => {
      // here we can interact with ydoc. 
      // but the problem is that the information about COMMENT POSITIONS are stored somewhere in the
      // tiptap annotation plugin state, and we are not able to get them.
    });

    this.editor = new Editor({
      extensions: [
        Document,
        CollaborationAnnotation.configure({
          document: this.ydoc,
          onUpdate: items => { this.comments = items },
          map: this.map,
          instance: 'editor1',
        }),
        Collaboration.configure({
          document: this.ydoc,
        }),
      ],
      content: `<p>Annotations can be used to add additional information to the content, for example comments. They live on a different level than the actual editor content. </p>`,
    })
  },
}
</script>
```

So the summarize this questions very basic in a first instance: How could we get all comments with the position to store them? Maybe a method `findAll()`? https://github.com/ueberdosis/tiptap/blob/main/docs/src/demos/Experiments/CollaborationAnnotation/extension/AnnotationState.ts#L31

Hopefully we could provide enough information of what is the problem. Thanks for taking time.

**Comment by hanspagel** [comment]
Hi @nadar! Thanks for sponsoring our open source work! 🙌

Positions are stored in the Y.Doc:

```js
CollaborationAnnotation.configure({
  document: this.ydoc,
})
```

The problem with that experiment is, there are major problems when splitting, deleting, merging nodes. Positions aren’t calculated correctly. That’s not an easy to solve problem.

I’ll try to set aside some time to add all the information I have, and list all known bugs on the page.

**Reply by nadar** [reply]
As we can see, the main task (to create an annotation for a given word) works perfectly, we could not encounter those bugs, we also do not need to have multiple collaborators in the project. we just want to add comments and display all comments next to the editor.

Could you please quickly elaborate how we could retrieve all comments (including the position) so we could store them in a database and retrieve them later (and load them again into the editor)? This would be very nice! As already mentioned, accessing the Ydoc's map with `this.map.observe(())` we where able to ready all posted comments/annotations - but the word position is stored in the annotation plugin - somewhere :smile: 

Thanks for your reply, we appreciate that!

**Reply by hanspagel** [reply]
Ah, now I got it! 

You don’t want to use it for collaborative editing? Than that should work just fine, but needs polishing (better API, docs, and so on). I’ll discuss this with Philipp next week and see what we can do to finish that one without the Collab part soon.

**Reply by DannyFeliz** [reply]
@nadar I'm with you, I'm needing this so bad

**Reply by nadar** [reply]
@hanspagel Yes exactly. As we knew we do not need the multi users/realtime collaboration stuff, we where trying to do without, just with a tiptap extension but we where not able to handle the multi selection option when 2 comments overlapping each other - so we decided to take a closer look at the annotation experiment, even it has that drawback of a technology (y.doc) we might not use it seems to work perfect. I think it "just" needs an option to retrieve all comments, whether they are selected or not including its position (to store in db) and a command to initialize them again (read from db). That would be awesome! If you need more informations, or a working codesandbox example or something which helps you guys - let me know, i'm here 🥳 

@DannyFeliz do you need to have multiple users working on the same doc or just the annotation/comment system like i do? Because for the multi users/colab system afaik there is still some more improvements need in other libraries, see [this issue comment](https://github.com/yjs/y-prosemirror/issues/44)

**Reply by evangow** [reply]
I am also looking into using TipTap for a similar use-case. I do not need the realtime collaboration parts to this annotation plugin. I'd just like a way for others to comment on a document that remains static after it's been published

**Reply by nadar** [reply]
@hanspagel is there any "news" on this?

**Reply by hanspagel** [reply]
Sorry, couldn’t find time to work on this. The good good news is, it became part of list of a few extensions we’d like to build. The bad news it’ll take some time until we start it, partly because we have a few weeks of summer holidays coming, but also because we try to keep the focus on stability.

**Reply by hanspagel** [reply]
Oh, and one more update: We had an idea last week to make that extension finally work with Y.js, and that approach would also work without Y.js (which would be great from the maintainer perspective, but also from the developer perspective). But I would need some time to test for different use cases before deciding to follow that approach.

**Reply by nadar** [reply]
Could you please contact me by email, we would like to make you an offer to fund this plugin - otherwise we probably need to look for other solutions.

**Reply by nadar** [reply]
@hanspagel not interested?

**Reply by DannyFeliz** [reply]
@philippkuehn 👀

**Comment by strdr4605** [comment]
Are there any updates on this? I could not find this plugin at https://tiptap.dev/extensions? Is it still experimental/WIP?

**Reply by hanspagel** [reply]
It’s available as a pro extension, but doesn’t have support for collab yet.

https://tiptap.dev/api/extensions/annotation
